feat: Add v13 -> v14 migration guide (#1011)
discordjs/guide: guide/additional-info/changes-in-v14.md

# Updating from v13 to v14

## Before you start

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

## Breaking Changes

### Enum Values

Any areas that used to accept a `string` or `number` type for an enum parameter will now only accept exclusively `number`s.

In addition, the old enums exported by discord.js v13 and lower are replaced with new enums from discord-api-types (link here).

<details>
<summary> New enum differences </summary>

>   Most of the difference between enums from discord.js and discord-api-type can be summarized as so:

Reviewer: typo
```suggestion
  Most of the difference between enums from discord.js and discord-api-types can be summarized as so:
```

1. Enums are singular ie `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`
2. Enums that are prefixed with `Message` no longer have the `Message` prefix ie `MessageButtonStyles` -> `ButtonStyle`
3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE` ie `.CHAT_INPUT` -> `.ChatInput`
 </details>

There are two recommended ways of representing enum values:

1. Use the actual enum type: `ButtonStyle.Primary`
2. Continue using a string representation but instead use the new `EnumResolver`:

```js
const { EnumResolver } = require('discord.js');
const enumValue = EnumResolvers.resolveButtonStyle('PRIMARY');
```

Areas like JSON slash commands and JSON message compononents will likely need to be modified to accomodate these changes:

**Application commands**:

```diff
+ const { ApplicationCommandType, ApplicationCommandOptionType } = require('discord.js');

const command = {
  name: 'ping',
- type: 'CHAT_INPUT',

> + type: ApplicationCommandType.ChatInput

Reviewer 2: ```suggestion
+ type: ApplicationCommandType.ChatInput,
```

  options: [
    name: 'option',
    description: 'A sample option'
-   type: 'STRING',
+   type: ApplicationCommandOptionType.String
  ],
};
```

**Buttons**:

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {
  label: 'test'

> - style: 'PRIMARY'

Reviewer 2: ```suggestion
- style: 'PRIMARY',
```

> + style: ButtonStyle.Primary

Reviewer 2: ```suggestion
+ style: ButtonStyle.Primary,
```

  customId: '1234'
}
```

### Event Removals

> `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.

Reviewer 2: ```suggestion
The `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.
```

> `applicationCommandCreate`, `applicationCommandDelete` and `applicationCommandUpdate` have all been removed. Refer to [this PR](https://github.com/discordjs/discord.js/pull/6492) for context.

Reviewer 2: I wouldn't shorten "pull request" down to "PR" when we're writing a guide. Keeping abbreviations to a minimum seems ideal to me? It's also inconsistent when you haven't abbreviated it below.

# Utility Changes

> `Util#removeMentions` has been removed.

Reviewer 2: I would explain the use of `allowedMentions` as a replacement. This utility method was in place due to mention injections - it would be a nice quick pointer as to what users should be using to control their mentions instead.

### CDN Changes

Methods that return CDN URLs will now return a dynamic image URL (if available). This behavior can be overriden by setting `forceStatic` to `false` in the `MakeURLOptions` parameters.

### Guild Object Changes

> All `Guild#setXPositions` methods have been removed, use `RoleManager#setPositions` and `GuildChannelManager#setPositions` instead.

Reviewer 2: Instead of saying:

> All `Guild#setXPositions` methods have been removed

I would actually list out what was removed. It feels strange you haven't since you've done it later for the replacements.

```suggestion
`Guild#setRolePositions` and `Guild#setChannelPositions` have been removed. Use `RoleManager#setPositions` and `GuildChannelManager#setPositions` instead respectively.
```

### Role Manager

> `#comparePositions` has been moved from `Role` to `RoleManager`

Reviewer 2: The style of this message doesn't match what's above it. In fact, it's the opposite, I would just stick to the same format as above for consistency (`Role#comparePositions` was removed. Use the method on the manager instead etc.).

> `#comparePositions` has been removed from `Role`. Use `RoleManager#comparePositions` instead.

Reviewer 2: Use `.` here instead of `#` as it's not an instance of the `Role` class (also added the name to match above):

```suggestion
`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.
```

Also, should this be `### Role Manager` or `### Role`?

# Updating from v13 to v14

## Before you start

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

## Breaking Changes

### Enum Values

Any areas that used to accept a `string` or `number` type for an enum parameter will now only accept exclusively `number`s.

In addition, the old enums exported by discord.js v13 and lower are replaced with new enums from discord-api-types (link here).

<details>
<summary> New enum differences </summary>
  Most of the difference between enums from discord.js and discord-api-types can be summarized as so:

> 1. Enums are singular ie `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`

Reviewer 3: ```suggestion
1. Enums are singular, i.e., `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`
```

> 2. Enums that are prefixed with `Message` no longer have the `Message` prefix ie `MessageButtonStyles` -> `ButtonStyle`

Reviewer 3: ```suggestion
2. Enums that are prefixed with `Message` no longer have the `Message` prefix, i.e., `MessageButtonStyles` -> `ButtonStyle`
```

> 3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE` ie `.CHAT_INPUT` -> `.ChatInput`

Reviewer 3: ```suggestion
3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE`, i.e., `.CHAT_INPUT` -> `.ChatInput`
```

const command = {
  name: 'ping',
- type: 'CHAT_INPUT',
+ type: ApplicationCommandType.ChatInput,
  options: [
    name: 'option',
    description: 'A sample option'
-   type: 'STRING',
+   type: ApplicationCommandOptionType.String
  ],
};
```

**Buttons**:

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {
  label: 'test'
- style: 'PRIMARY',
+ style: ButtonStyle.Primary,
  customId: '1234'
}
```

### Event Removals

The `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.

`applicationCommandCreate`, `applicationCommandDelete` and `applicationCommandUpdate` have all been removed. Refer to [this pull request](https://github.com/discordjs/discord.js/pull/6492) for context.

### REST Events

The following discord.js events:
- `invalidRequestWarning`
- `request`
- `response`
- `rateLimited`
- `newListener`
- `removeListener`

Have been removed from the `Client` in discord.js. Instead you should access these events from `Client#rest`.

### Utility Changes

`Util#removeMentions` has been removed, to control mentions you should use `allowedMentions` on `MessageOptions` instead.

### CDN Changes

> Methods that return CDN URLs will now return a dynamic image URL (if available). This behavior can be overriden by setting `forceStatic` to `false` in the `MakeURLOptions` parameters.

Reviewer 4: ```suggestion
Methods that return CDN URLs will now return a dynamic image URL (if available). This behavior can be overriden by setting `forceStatic` to `true` in the `MakeURLOptions` parameters.
```

### Guild Object Changes

`Guild#setRolePositions` and `Guild#setChannelPositions` have been removed. Use `RoleManager#setPositions` and `GuildChannelManager#setPositions` instead respectively.

### Thread Member Manager

`ThreadMemberManager#fetch` now only takes a single object of type `ThreadMemberFetchOptions`.

### Role Manager

`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.

### `.deleted` Field(s) have been removed

You can no longer use `#deleted` to check if a structure was deleted or not. 

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### Channel changes

`Channel#createAt` is now nullable. On any regular channel `#createAt` or private thread this value will always be non-null. This value is only nullable for public threads. 

::: tip
TypeScript users should narrow `Channel` types via type guards in order to get more specific typings.
:::

### Message Embed Changes

- `MessageEmbed` has now been renamed to `Embed`

> - `#setFooter` now accepts a sole object `APIEmbedFooter` object. (add link to dapi site)

Reviewer 4: #setFooter takes FooterOptions which accepts camelCase not snake_case

https://github.com/discordjs/discord.js/blob/00ce1c56ac224691a8691a3525cb14ae002319c6/packages/builders/src/messages/embed/Embed.ts#L234

### Message Component Changes

MessageComponents have been renamed as well. They no longer have the `Message` prefix:

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

### Interaction typeguard changes

The following typeguards on `Interaction` have been renamed:

```diff
- interaction.isCommand()
+ interaction.isChatInputCommand()

- interaction.isContextMenu()
+ interaction.isContextMenuCommand()
```

In addition, `Interaction#isCommand`, now indicates whether the command is an *application command* or not. This differs from the previous implementation where `isCommand()` indicated if the interaction was a chat input command or not.

### Voices Region Changes

`VoiceRegion#vip` has been removed as the field is no longer part of the API.

### Application Changes

> `Application#fetchAssets` has been removed as the used for this method is no longer available from the API.

Reviewer 4: ```suggestion
`Application#fetchAssets` has been removed as the use for this method is no longer available from the API.
```

Reviewer 2: It would be more accurate to state:

> X has been removed as it is no longer supported by the API.

Clear and concise, whereas "the use for this method is no longer available" seems like it opens up avenues that questions its removal.

const enumValue = EnumResolvers.resolveButtonStyle('PRIMARY');
```

Areas like JSON slash commands and JSON message components will likely need to be modified to accommodate these changes:

#### Common Application Command Data changes

```diff
+ const { ApplicationCommandType, ApplicationCommandOptionType } = require('discord.js');

const command = {
  name: 'ping',
- type: 'CHAT_INPUT',
+ type: ApplicationCommandType.ChatInput,
  options: [
    name: 'option',
    description: 'A sample option'
-   type: 'STRING',
+   type: ApplicationCommandOptionType.String
  ],
};
```

#### Common Button Data changes

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {
  label: 'test'
- style: 'PRIMARY',
+ style: ButtonStyle.Primary,
  customId: '1234'
}
```

### Events

The `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.

`applicationCommandCreate`, `applicationCommandDelete` and `applicationCommandUpdate` have all been removed. Refer to [this pull request](https://github.com/discordjs/discord.js/pull/6492) for context.

#### REST Events

The following discord.js events:
- `invalidRequestWarning`
- `request`

> - `response`

Reviewer 5: these were named differently in djs

- `rateLimited`
- `newListener`

> - `removeListener`

Reviewer 5: don't list these

# Updating from v13 to v14

## Before you start

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

## Breaking Changes

### Common Breakages

> ### Enum Values

Reviewer 4: ```suggestion
#### Enum Values
```

Author: I'll keep it as-is for now

<summary> New enum differences </summary>
  Most of the difference between enums from discord.js and discord-api-types can be summarized as so:

1. Enums are singular, i.e., `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`
2. Enums that are prefixed with `Message` no longer have the `Message` prefix, i.e., `MessageButtonStyles` -> `ButtonStyle`
3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE`, i.e., `.CHAT_INPUT` -> `.ChatInput`
 </details>

There are two recommended ways of representing enum values:

1. Use the actual enum type: `ButtonStyle.Primary`
2. Continue using a string representation but instead use the new `EnumResolver`:

```js
const { EnumResolver } = require('discord.js');
const enumValue = EnumResolvers.resolveButtonStyle('PRIMARY');
```

Areas like JSON slash commands and JSON message components will likely need to be modified to accommodate these changes:

#### Common Application Command Data changes

```diff
+ const { ApplicationCommandType, ApplicationCommandOptionType } = require('discord.js');

const command = {
  name: 'ping',
- type: 'CHAT_INPUT',
+ type: ApplicationCommandType.ChatInput,
  options: [
    name: 'option',

>     description: 'A sample option'

Reviewer 6: ```suggestion
    description: 'A sample option',
```

-   type: 'STRING',
+   type: ApplicationCommandOptionType.String
  ],
};
```

#### Common Button Data changes

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {

>   label: 'test'

Reviewer 6: ```suggestion
  label: 'test',
```

### Interaction

The following typeguards on `Interaction` have been renamed:

```diff
- interaction.isCommand()
+ interaction.isChatInputCommand()

- interaction.isContextMenu()
+ interaction.isContextMenuCommand()
```

In addition, `#isCommand`, now indicates whether the command is an *application command* or not. This differs from the previous implementation where `#isCommand` indicated if the interaction was a chat input command or not.

### MessageComponent

MessageComponents have been renamed as well. They no longer have the `Message` prefix:

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

### MessageEmbed

- `MessageEmbed` has now been renamed to `Embed`

- `#setFooter` now accepts a sole `FooterOptions` object. (add link to dapi site)

- `#addField` and `#addFields` both accept an object or array of `APIEmbedField`(s) respectively. (add link to dapi site)

> #### REST Events

Reviewer 4: ```suggestion
### REST Events
```

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

## Breaking Changes

### Common Breakages

### Enum Values

Any areas that used to accept a `string` or `number` type for an enum parameter will now only accept exclusively `number`s.

In addition, the old enums exported by discord.js v13 and lower are replaced with new enums from discord-api-types (link here).

<details>
<summary> New enum differences </summary>
  Most of the difference between enums from discord.js and discord-api-types can be summarized as so:

1. Enums are singular, i.e., `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`
2. Enums that are prefixed with `Message` no longer have the `Message` prefix, i.e., `MessageButtonStyles` -> `ButtonStyle`
3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE`, i.e., `.CHAT_INPUT` -> `.ChatInput`
 </details>

There are two recommended ways of representing enum values:

1. Use the actual enum type: `ButtonStyle.Primary`

> 2. Continue using a string representation but instead use the new `EnumResolver`:

Reviewer 7: ```suggestion
2. Continue using a string representation but instead use the new `EnumResolvers`:
```

```js

> const { EnumResolver } = require('discord.js');
> const enumValue = EnumResolvers.resolveButtonStyle('PRIMARY');

Reviewer 7: ```suggestion
const { EnumResolvers } = require('discord.js');
const enumValue = EnumResolvers.resolveButtonStyle('PRIMARY');
```

```

### RoleManager

`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.

### ThreadMemberManager

`#fetch` now only takes a single object of type `ThreadMemberFetchOptions`.

### Util

`#removeMentions` has been removed, to control mentions you should use `allowedMentions` on `MessageOptions` instead.

### `.deleted` Field(s) have been removed

You can no longer use `#deleted` to check if a structure was deleted or not. 

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### VoiceChannel

`#editable` has been removed, instead you should use `GuildChannel#manageable` for checking this permission.

Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

### VoicesRegion Changes

`#vip` has been removed as the field is no longer part of the API.

### Webhook

`#fetchMessage` now only takes on sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

New typeguards have been added:
- `#isText`*
- `#isDM`*
- `#isGroupDM`
- `#isVoice`*
- `#isVoiceBased`
- `#isTextBased`
- `#isStage`
- `#isCategory`
- `#isNews`
- `#isStore`

> *These methods existed previously but have different behaviors refer to the docs for their specific changes.

Reviewer 8: ```suggestion
*These methods existed previously but perform different behaviours. Refer to the docs for their specific changes.
```

Grammar issue

# Updating from v13 to v14

## Before you start

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

## Breaking Changes

### Common Breakages

### Enum Values

Any areas that used to accept a `string` or `number` type for an enum parameter will now only accept exclusively `number`s.

In addition, the old enums exported by discord.js v13 and lower are replaced with new enums from discord-api-types (link here).

<details>
<summary> New enum differences </summary>
  Most of the difference between enums from discord.js and discord-api-types can be summarized as so:

1. Enums are singular, i.e., `ApplicationCommandOptionTypes` -> `ApplicationCommandOptionType`
2. Enums that are prefixed with `Message` no longer have the `Message` prefix, i.e., `MessageButtonStyles` -> `ButtonStyle`
3. Enum values are `PascalCase` rather than `SCREAMING_SNAKE_CASE`, i.e., `.CHAT_INPUT` -> `.ChatInput`

>  </details>

Reviewer 8: ```suggestion
</details>
```

#### Common Client Initialization Changes

```diff
- const { Client, Intents } = require('discord.js');
+ const { Client, GatewayIntentBits, Partials } = require('discord.js');

- const client = new Client({ intents: [Intents.FLAGS.GUILDS], partials: ['CHANNEL'] });
+ const client = new Client({ intents: [GatewayIntentBits.Guilds], partials: [Partials.Channel] });
```

#### Common Application Command Data changes

```diff
+ const { ApplicationCommandType, ApplicationCommandOptionType } = require('discord.js');

const command = {
  name: 'ping',
- type: 'CHAT_INPUT',
+ type: ApplicationCommandType.ChatInput,
  options: [
    name: 'option',
    description: 'A sample option',
-   type: 'STRING',
+   type: ApplicationCommandOptionType.String
  ],
};
```

#### Common Button Data changes

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {
  label: 'test',
- style: 'PRIMARY',
+ style: ButtonStyle.Primary,
  customId: '1234'
}
```

### Application

`#fetchAssets` has been removed as it is no longer supported by the API.

### BitField

> - BitField constituents now have a `BitField` suffix to avoid naming conflicts with the enum names:

Reviewer 8: ```suggestion
- BitField constituents now have a `BitField` suffix to avoid naming conflicts with the enum names:

```

```diff
- new Permissions()
+ new PermissionsBitField()

- new MessageFlags()
+ new MessageFlagsBitField()

- new ThreadMemberFlags()
+ new ThreadMemberFlagsBitField()

- new UserFlags()
+ new UserFlagsBitField()

- new SystemChannelFlags()
+ new SystemChannelFlagsBitField()

- new ApplicationFlags()
+ new ApplicationFlagsBitField()

- new Intents()

> + new IntentsBitField

Reviewer 9: ```suggestion
+ new IntentsBitField()
 ```

- new ActivityFlags()
+ new ActivityFlagsBitField()
```

- `#flags` has been renamed to `#Flags`

### CDN

- Methods that return CDN URLs will now return a dynamic image URL (if available). This behavior can be overridden by setting `forceStatic` to `true` in the `MakeURLOptions` parameters.
- CDN methods now default to fetching `.webp` assets (if available), instead of `.png` assets.

### Channel

> - `#createdAt` and `#createdTimestamp` are now nullable. On any regular channel or private thread `#createdAt` and `#createTimestamp` will always be non-null. This value is only nullable for public threads. 

Reviewer 8: ```suggestion
- `#createdAt` and `#createdTimestamp` are now nullable. On any regular channel or private thread `#createdAt` and `#createTimestamp` will always be non-null. This value is only nullable for public threads.
```

- `#isText` has been renamed to `#isTextBased`

> - `#isVoice` has been renamed to `#isVoiceBased` 

Reviewer 8: ```suggestion
- `#isVoice` has been renamed to `#isVoiceBased`
```

::: tip
TypeScript users should narrow `Channel` types via type guards in order to get more specific typings.
:::

### CommandInteractionOptionResolver

`#getMember` no longer has a parameter for `required`, check out [this pull request](https://github.com/discordjs/discord.js/pull/7188) for details.

### Events

The `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.

`applicationCommandCreate`, `applicationCommandDelete` and `applicationCommandUpdate` have all been removed. Refer to [this pull request](https://github.com/discordjs/discord.js/pull/6492) for context.

### Guild

`#setRolePositions` and `#setChannelPositions` have been removed. Use `RoleManager#setPositions` and `GuildChannelManager#setPositions` instead respectively.

### Interaction

The following typeguards on `Interaction` have been renamed:

```diff
- interaction.isCommand()
+ interaction.isChatInputCommand()

- interaction.isContextMenu()
+ interaction.isContextMenuCommand()
```

> In addition, `#isCommand`, now indicates whether the command is an *application command* or not. This differs from the previous implementation where `#isCommand` indicated if the interaction was a chat input command or not.

Reviewer 8: ```suggestion
In addition, `#isCommand`, now indicates whether the command is an _application command_ or not. This differs from the previous implementation where `#isCommand` indicated if the interaction was a chat input command or not.
```

### Invite

`#channel` and `#inviter` are now getters and resolve structures from the cache.

### MessageComponent

MessageComponents have been renamed as well. They no longer have the `Message` prefix:

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

### MessageEmbed

- `MessageEmbed` has now been renamed to `Embed`

> 

Reviewer 9: ```suggestion

- `#setAuthor` now accepts a sole `AuthorOptions` object. (add link to dapi site)

```

- `#setFooter` now accepts a sole `FooterOptions` object. (add link to dapi site)

- `#addField` and `#addFields` both accept an object or array of `APIEmbedField`(s) respectively. (add link to dapi site)

### PartialTypes

The `PartialTypes` string array has been removed, instead use the `Partials` enum.

### Permissions

Thread permissions `USE_PUBLIC_THREADS` and `USE_PRIVATE_THREADS` have been removed as they are now deprecated in the API. Instead use the newer `*Threads` permission flags.

### PermissionOverwritesManager

Overwrites are now keyed by the `PascalCase` permission key rather than the `SCREAMING_SNAKE_CASE` permission key.

### REST Events

> The following discord.js events:

Reviewer 8: ```suggestion
The following discord.js events:

```

- `invalidRequestWarning`
- `apiRequest`
- `apiResponse`
- `rateLimit`

Have been removed from the `Client` in discord.js. Instead you should access these events from `Client#rest`. In addition `apiRequest`, `apiResponse` and `rateLimit` events have been renamed:

```diff
- client.on('apiRequest', ...);
+ client.rest.on('request', ...);

- client.on('apiResponse', ...);
+ client.rest.on('response', ...);

- client.on('rateLimit', ...);
+ client.rest.on('rateLimited', ...);
```

### RoleManager

`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.

### ThreadMemberManager

`#fetch` now only takes a single object of type `ThreadMemberFetchOptions`.

### Util

`#removeMentions` has been removed, to control mentions you should use `allowedMentions` on `MessageOptions` instead.

> 

Reviewer 8: ```suggestion
```

### `.deleted` Field(s) have been removed

> You can no longer use `#deleted` to check if a structure was deleted or not. 

Reviewer 8: ```suggestion
You can no longer use `#deleted` to check if a structure was deleted or not.
```

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### VoiceChannel

`#editable` has been removed, instead you should use `GuildChannel#manageable` for checking this permission.

Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

### VoicesRegion Changes

`#vip` has been removed as the field is no longer part of the API.

### Webhook

`#fetchMessage` now only takes on sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

New typeguards have been added:
- `#isText`*
- `#isDM`*
- `#isGroupDM`
- `#isVoice`*
- `#isVoiceBased`

> - `#isTextBased`

Reviewer 8: ```suggestion
- `#isTextBased`
- `#isDMBased`
```

See https://github.com/discordjs/discord.js/pull/7362 (`.isDMBased()` has been added)

PS. I think this should be in alphabetical order?

> New typeguards have been added:
> - `#isText`*
> - `#isDM`*
> - `#isGroupDM`
> - `#isVoice`*
> - `#isVoiceBased`
> - `#isTextBased`
> - `#isStage`
> - `#isCategory`
> - `#isNews`
> - `#isStore`
> 
> *These methods existed previously but behaved differently. Refer to the docs for their specific changes.

Reviewer 8: ```suggestion
New typeguards have been added:

- `#isText`\*
- `#isDM`\*
- `#isGroupDM`
- `#isVoice`\*
- `#isVoiceBased`
- `#isTextBased`
- `#isStage`
- `#isCategory`
- `#isNews`
- `#isStore`

\*These methods existed previously but behaved differently. Refer to the docs for their specific changes.
```

> ### Enum Resolvers

Reviewer 8: ```suggestion
### Enum Resolvers

```

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

### MessageEmbed

- `MessageEmbed` has now been renamed to `Embed`

- `#setAuthor` now accepts a sole `AuthorOptions` object. (add link to dapi site)

- `#setFooter` now accepts a sole `FooterOptions` object. (add link to dapi site)

- `#addField` and `#addFields` both accept an object or array of `APIEmbedField`(s) respectively. (add link to dapi site)

### PartialTypes

The `PartialTypes` string array has been removed, instead use the `Partials` enum.

### Permissions

Thread permissions `USE_PUBLIC_THREADS` and `USE_PRIVATE_THREADS` have been removed as they are now deprecated in the API. Instead use the newer `*Threads` permission flags.

### PermissionOverwritesManager

Overwrites are now keyed by the `PascalCase` permission key rather than the `SCREAMING_SNAKE_CASE` permission key.

### REST Events

The following discord.js events:

- `invalidRequestWarning`
- `apiRequest`
- `apiResponse`
- `rateLimit`

> Have been removed from the `Client` in discord.js. Instead you should access these events from `Client#rest`. In addition `apiRequest`, `apiResponse` and `rateLimit` events have been renamed:

Reviewer 10: ```suggestion
Have been removed from the `Client` in discord.js. Instead you should access these events from `Client#rest`. In addition, the `apiRequest`, `apiResponse` and `rateLimit` events have been renamed:
```

### `.deleted` Field(s) have been removed

You can no longer use `#deleted` to check if a structure was deleted or not.

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### VoiceChannel

`#editable` has been removed, instead you should use `GuildChannel#manageable` for checking this permission.

Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

### VoicesRegion Changes

`#vip` has been removed as the field is no longer part of the API.

### Webhook

`#fetchMessage` now only takes on sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

New typeguards have been added:

- `#isText`\*
- `#isDM`\*
- `#isGroupDM`
- `#isVoice`\*
- `#isVoiceBased`
- `#isTextBased`
- `#isDMBased`
- `#isStage`
- `#isCategory`
- `#isNews`
- `#isStore`

\*These methods existed previously but behaved differently. Refer to the docs for their specific changes.

### Enum Resolvers

The new `EnumResolvers` class allows you to transform `SCREAMING_SNAKE_CASE` enum keys to an enum value.

```js
const { EnumResolvers } = require('discord.js');

// Returns `ButtonStyle.Primary`

> const buttonStyle = EnumResolvers.buttonStyle('PRIMARY');

Reviewer 11: ```suggestion
const buttonStyle = EnumResolvers.resolveButtonStyle('PRIMARY');
```

### Interaction

The following typeguards on `Interaction` have been renamed:

```diff
- interaction.isCommand()
+ interaction.isChatInputCommand()

- interaction.isContextMenu()
+ interaction.isContextMenuCommand()
```

In addition, `#isCommand`, now indicates whether the command is an _application command_ or not. This differs from the previous implementation where `#isCommand` indicated if the interaction was a chat input command or not.

### Invite

`#channel` and `#inviter` are now getters and resolve structures from the cache.

### MessageComponent

MessageComponents have been renamed as well. They no longer have the `Message` prefix:

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

### MessageEmbed

- `MessageEmbed` has now been renamed to `Embed`.

- `#setAuthor` now accepts a sole `AuthorOptions` object. (add link to dapi site)

- `#setFooter` now accepts a sole `FooterOptions` object. (add link to dapi site)

> - `#addField` and `#addFields` both accept an object or array of `APIEmbedField`(s) respectively. (add link to dapi site)

Reviewer 4: Breaking: addField is now removed.

- new ApplicationFlags()
+ new ApplicationFlagsBitField()

- new Intents()
+ new IntentsBitField()

- new ActivityFlags()
+ new ActivityFlagsBitField()
```

- `#FLAGS` has been renamed to `#Flags`

### CDN

Methods that return CDN URLs will now return a dynamic image URL (if available). This behavior can be overridden by setting `forceStatic` to `true` in the `ImageURLOptions` parameters.

### Channel

- `#createdAt` and `#createdTimestamp` are now nullable. On any regular channel or private thread `#createdAt` and `#createTimestamp` will always be non-null. This value is only nullable for public threads.
- `#isText` has been renamed to `#isTextBased`
- `#isVoice` has been renamed to `#isVoiceBased`

::: tip
TypeScript users should narrow `Channel` types via type guards in order to get more specific typings.
:::

### CommandInteractionOptionResolver

`#getMember` no longer has a parameter for `required`, check out [this pull request](https://github.com/discordjs/discord.js/pull/7188) for details.

### `Constants`

- Many constant objects and key arrays are now top-level exports for example:

```diff
- const { Constants } = require('discord.js');
- const { Colors } = Constants;

> + const { Colors } = require('discord.js')

Reviewer 3: ```suggestion
+ const { Colors } = require('discord.js');
```

### Guild

`#setRolePositions` and `#setChannelPositions` have been removed. Use `RoleManager#setPositions` and `GuildChannelManager#setPositions` instead respectively.

### Interaction

The following typeguards on `Interaction` have been renamed:

```diff
- interaction.isCommand()
+ interaction.isChatInputCommand()

- interaction.isContextMenu()
+ interaction.isContextMenuCommand()
```

In addition, `#isCommand`, now indicates whether the command is an _application command_ or not. This differs from the previous implementation where `#isCommand` indicated if the interaction was a chat input command or not.

### Invite

`#channel` and `#inviter` are now getters and resolve structures from the cache.

### MessageComponent

MessageComponents have been renamed as well. They no longer have the `Message` prefix:

```diff
- const button = new MessageButton();
+ const button = new ButtonComponent();

- const selectMenu = new MessageSelectMenu();
+ const selectMenu = new SelectMenuComponent();

- const actionRow = new MessageActionRow();
+ const actionRow = new ActionRow();
```

### MessageEmbed

> - `MessageEmbed` has now been renamed to `Embed`

Reviewer 3: ```suggestion
- `MessageEmbed` has now been renamed to `Embed`.
```

- `apiRequest`
- `apiResponse`
- `invalidRequestWarning`
- `rateLimit`

Have been removed from the `Client` in discord.js. Instead you should access these events from `Client#rest`. In addition, the `apiRequest`, `apiResponse` and `rateLimit` events have been renamed:

```diff
- client.on('apiRequest', ...);
+ client.rest.on('request', ...);

- client.on('apiResponse', ...);
+ client.rest.on('response', ...);

- client.on('rateLimit', ...);
+ client.rest.on('rateLimited', ...);
```

### RoleManager

`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.

### ThreadMemberManager

`#fetch` now only takes a single object of type `ThreadMemberFetchOptions`.

### Util

`#removeMentions` has been removed, to control mentions you should use `allowedMentions` on `MessageOptions` instead.

### `.deleted` Field(s) have been removed

You can no longer use `#deleted` to check if a structure was deleted or not.

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### VoiceChannel

`#editable` has been removed, instead you should use `GuildChannel#manageable` for checking this permission.

> Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

Reviewer 3: ```suggestion
Many of the analogous enums can be found in the discord-api-types docs. (link website here)
```

### VoiceRegion

`#vip` has been removed as the field is no longer part of the API.

### Webhook

`#fetchMessage` now only takes one sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

New typeguards have been added:

- `#isCategory`

> - `#isDM`\*

Reviewer 9: ```suggestion
- `#isDM`
```
Seeing the docs for stable, isDM wasn't a thing back then

- `#isDMBased`
- `#isGroupDM`
- `#isNews`
- `#isStage`
- `#isStore`
- `#isText`\*
- `#isTextBased`
- `#isVoice`\*
- `#isVoiceBased`

\*These methods existed previously but behaved differently. Refer to the docs for their specific changes.

### Collection

- Added `#merge` and `#combineEntries` methods.

> - New type: `ReadonlyCollection` which indicates an immutable `Collection`

Reviewer 3: ```suggestion
- New type: `ReadonlyCollection` which indicates an immutable `Collection`.
```

```js
const { EnumResolvers } = require('discord.js');

// Returns `ButtonStyle.Primary`
const buttonStyle = EnumResolvers.resolveButtonStyle('PRIMARY');
```
### GuildEmojiManager

Added `#delete` and `#edit` methods for managing existing guild emojis.

### GuildPreview

Added `#stickers` property.

### GuildScheduledEvent

Added `#coverImage` and `coverImageURL` for getting cover image data for a guild scheduled event.

### Interaction

Added `#isRepliable` to check whether a given interaction can be replied to.

### MessageOptions

You can now send `MessageFlags` via the `#flags` property on `MessageOptions`.

### RichPresenceAssets

`RichPresenceAssets` now contains support for Youtube content and custom images.

### Thread

Added `#createdTimestamp` which indicates when the thread was created.

### Thread create event

The `threadCreate` event now includes a new parameter: `newlyCreated`. This is a boolean that indicates whether the thread was just created or not.

### Unsafe Builders

Unsafe builders operate exactly like regular builders except they perform no validation on input. Unsafe builders are named by adding an `Unsafe` prefix to a regular builder.

### Webhook

> Added `applicationID` property.

Reviewer 9: ```suggestion
Added `applicationId` property.
```

::: tip
TypeScript users should narrow `Channel` types via type guards in order to get more specific typings.
:::

### CommandInteractionOptionResolver

`#getMember` no longer has a parameter for `required`, check out [this pull request](https://github.com/discordjs/discord.js/pull/7188) for details.

### `Constants`

- Many constant objects and key arrays are now top-level exports for example:

```diff
- const { Constants } = require('discord.js');
- const { Colors } = Constants;
+ const { Colors } = require('discord.js');
```

- The refactored constants structures have `PascalCase` member names as opposed to `SCREAMING_SNAKE_CASE` member names.

- Many of the exported constants structures have been replaced and renamed:
```diff
- Opcodes
+ GatewayOpcodes

- WSEvents
+ GatewayDispatchEvents

- WSCodes
+ GatewayCloseCodes

- InviteScopes
+ OAuth2Scopes
```

### Events

The `message` and `interaction` events are now removed. Use `messageCreate` and `interactionCreate` instead.

`applicationCommandCreate`, `applicationCommandDelete` and `applicationCommandUpdate` have all been removed. Refer to [this pull request](https://github.com/discordjs/discord.js/pull/6492) for context.

> The `ThreadMembersUpdate` event now only emits the thread, the users who were added & and users who were removed respectively.

Reviewer 4: ```suggestion
The `ThreadMembersUpdate` event now emits the thread, the users who were added & and users who were removed respectively.
```

ig only isn't needed.

Reviewer 4: I missed 
```diff
- & and
+ and
```

```

- Components received from the API are no longer directly mutable. If you wish to mutate a component from the API, use `ComponentBuilder#from`. For example, if you want to make a button mutable:

```diff
- const editedButton = receivedButton
-   .setDisabled(true);

+ import { ButtonBuilder } from 'discord.js';
+ const editedButton = ButtonBuilder.from(receivedButton)
+   .setDisabled(true);
```

### MessageSelectMenu

- `MessageSelectMenu` has been renamed to `SelectMenuBuilder`

- `SelectMenuBuilder#addOption()` has been removed. Use `SelectMenuBuilder#addOptions()` instead.

### MessageEmbed

- `MessageEmbed` has now been renamed to `EmbedBuilder`.

- `EmbedBuilder#setAuthor()` now accepts a sole `AuthorOptions` object. (add link to dapi site)

- `EmbedBuilder#setFooter()` now accepts a sole `FooterOptions` object. (add link to dapi site)

- `EmbedBuilder#addField()` has been removed. Use `EmbedBuilder#addFields()` instead.

- `EmbedBuilder#addFields()` accepts an object or a rest array of `APIEmbedField`(s): (add link to dapi site)

```diff
- new MessageEmbed().addFields([
-  { name: 'one', value: 'one' },
-  { name: 'two', value: 'two' },
- ]);

> + new Embed().addFields(

Reviewer 11: ```suggestion
+ new EmbedBuilder().addFields(
```

### Util

`Util.removeMentions()` has been removed. To control mentions, you should use `allowedMentions` on `MessageOptions` instead.

`Util.splitMessage()` has been removed. This utility method is something the developer themselves should do.

### `.deleted` Field(s) have been removed

You can no longer use the `deleted` property to check if a structure was deleted. See [this issue](https://github.com/discordjs/discord.js/issues/7091) for more information.

### VoiceChannel

`VoiceChannel#editable` has been removed. You should use `GuildChannel#manageable` instead.

Many of the analogous enums can be found in the discord-api-types docs. [discord-api-types](https://discord-api-types.dev/api/discord-api-types-v10/enum/ActivityFlags)

### VoiceRegion

`VoiceRegion#vip` has been removed as it is no longer part of the API.

### Webhook

`Webhook#fetchMessage()` now only takes one sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

Store channels have been removed as they are no longer part of the API.

`Channel#url` has been added which is a link to a channel, just like in the client.

Additionally, new typeguards have been added:

> -   `Channel#isCategory()`
> -   `Channel#isDM()`
> -   `Channel#isDMBased()`
> -   `Channel#isGroupDM()`
> -   `Channel#isNews()`
> -   `Channel#isStage()`
> -   `Channel#isText()`\*
> -   `Channel#isTextBased()`
> -   `Channel#isVoice()`\*
> -   `Channel#isVoiceBased()`
> 
> \*These methods existed previously but behaved differently. Refer to the docs for their specific changes.

Reviewer 4: ```suggestion
-   `Channel#isTextBased()`
-   `Channel#isVoiceBased()`
```

Reviewer 9: note that isDMBased still exists

### `.deleted` Field(s) have been removed

You can no longer use the `deleted` property to check if a structure was deleted. See [this issue](https://github.com/discordjs/discord.js/issues/7091) for more information.

### VoiceChannel

`VoiceChannel#editable` has been removed. You should use `GuildChannel#manageable` instead.

Many of the analogous enums can be found in the discord-api-types docs. [discord-api-types](https://discord-api-types.dev/api/discord-api-types-v10/enum/ActivityFlags)

### VoiceRegion

`VoiceRegion#vip` has been removed as it is no longer part of the API.

### Webhook

`Webhook#fetchMessage()` now only takes one sole object of type `WebhookFetchMessageOptions`.

## Features

### Channel

Store channels have been removed as they are no longer part of the API.

`Channel#url` has been added which is a link to a channel, just like in the client.

Additionally, new typeguards have been added:

-   `Channel#isDMBased()`
-   `Channel#isTextBased()`
-   `Channel#isVoiceBased()`

### Collection

-   Added `Collection#merge()` and `Collection#combineEntries()`.
-   New type: `ReadonlyCollection` which indicates an immutable `Collection`.

### Collector

> A new `ignore` has been added which is emitted whenever an element is not collected by the collector.

Reviewer 7: ```suggestion
A new `ignore` event has been added which is emitted whenever an element is not collected by the collector.
```

  options: [
    name: 'option',
    description: 'A sample option',
-   type: 'STRING',
+   type: ApplicationCommandOptionType.String,
  ],
};
```

##### Common Button Data changes

```diff
+ const { ButtonStyle } = require('discord.js');

const button = {
  label: 'test',
- style: 'PRIMARY',
+ style: ButtonStyle.Primary,
  customId: '1234'
}
```

### Removal of method-based type guards

#### Channels

Some channel type guard methods that narrowed to one channel type have been removed. Instead compare the `type` property against a [ChannelType](https://discord-api-types.dev/api/discord-api-types-v10/enum/ChannelType) enum member to narrow channels.

```diff
-channel.isText()
+channel.type === ChannelType.GuildText

-channel.isVoice()
+channel.type === ChannelType.GuildVoice

-channel.isDM()
+channel.type === ChannelType.DM
```

#### Interactions

Similarly to channels, some interaction type guards have been removed, and replaced with `type` checks a [InteractionType](https://discord-api-types.dev/api/discord-api-types-v10/enum/InteractionType) enum member.

```diff
-interaction.isCommand();
+interaction.type === InteractionType.ApplicationCommand;

-interaction.isAutocomplete();

> +interaction.type === InteractionType.Autocomplete;

Reviewer 2: ```suggestion
+interaction.type === InteractionType.ApplicationCommandAutocomplete;
```

v14 requires Node 16.9 or higher to use, so make sure you're up to date. To check your Node version, use `node -v` in your terminal or command prompt, and if it's not high enough, update it! There are many resources online to help you with this step based on your host system.

### Builders are now included in v14

If you previously had `@discordjs/builders` manually installed it's *highly* recommended that you uninstall the package to avoid package naming conflicts.

**NPM**

> ```

Reviewer 4: ```suggestion
```bash
```

`#fetchAssets` has been removed as it is no longer supported by the API.

### BitField

- BitField constituents now have a `BitField` suffix to avoid naming conflicts with the enum names:
```diff
- new Permissions()
+ new PermissionsBitField()

- new MessageFlags()
+ new MessageFlagsBitField()

- new ThreadMemberFlags()
+ new ThreadMemberFlagsBitField()

- new UserFlags()
+ new UserFlagsBitField()

- new SystemChannelFlags()
+ new SystemChannelFlagsBitField()

- new ApplicationFlags()
+ new ApplicationFlagsBitField()

- new Intents()
+ new IntentsBitField

- new ActivityFlags()
+ new ActivityFlagsBitField()
```

> - `#flags` has been renamed to `#Flags`

Reviewer 9: ```suggestion
- `#FLAGS` has been renamed to `#Flags`
```

+ client.rest.on('response', ...);

- client.on('rateLimit', ...);
+ client.rest.on('rateLimited', ...);
```

### RoleManager

`Role.comparePositions` has been removed. Use `RoleManager#comparePositions` instead.

### ThreadMemberManager

`#fetch` now only takes a single object of type `ThreadMemberFetchOptions`.

### Util

`#removeMentions` has been removed, to control mentions you should use `allowedMentions` on `MessageOptions` instead.

### `.deleted` Field(s) have been removed

You can no longer use `#deleted` to check if a structure was deleted or not. 

Check out [the issue ticket](https://github.com/discordjs/discord.js/issues/7091) for more context.

### VoiceChannel

`#editable` has been removed, instead you should use `GuildChannel#manageable` for checking this permission.

Many of the analogous enums can be found be found in the discord-api-types docs (link website here)

### VoicesRegion Changes

`#vip` has been removed as the field is no longer part of the API.

## Features

### Channel

New typeguards have been added:
- `#isText`*
- `#isDM`*
- `#isGroupDM`
- `#isVoice`*
- `#isVoiceBased`
- `#isTextBased`
- `#isStage`
- `#isCategory`
- `#isNews`
- `#isStore`

> *These methods existed previously but have different behaviors refer to the docs for their specific changes.

Reviewer 4: Iirc only `isText` and `isVoice` these 2 methods existed previously which are renamed to `isTextBased` and `isVoiceBased` so ig adding these 2 change in breaking changes is worth it.
